torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -505,34 +531,56 @@ void RRefContext::addForkOfOwner(const RRefId& rrefId, const ForkId& forkId) {
   rrefForks.insert(forkId);
 }
 
+void RRefContext::addForkOfOwnerIfNotPresent(
+    const RRefId& rrefId,
+    const ForkId& forkId) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto& rrefForks = forks_[rrefId];
+  // We first check whether the child exists in rrefForks. It's possible
+  // the child may have been added by a previous send attempt, and this check
+  // (as opposed to an assertion here) ensures that messages that trigger this
+  // function are idempotent.
+  if (rrefForks.find(forkId) == rrefForks.end()) {
+    rrefForks.insert(forkId);
+  } else {
+    LOG(INFO) << "Ignoring duplicate request to add Fork of OwnerRRef with "
+              << "RRefId = " << rrefId << ", ForkId = " << forkId;
+  }
+}
+
 c10::intrusive_ptr<RRef> RRefContext::delForkOfOwner(
     const RRefId& rrefId,
     const ForkId& forkId) {
   c10::intrusive_ptr<RRef> deletedRRef;
   bool ownerReduced = false;
+  // There were previously multiple TORCH_CHECKs in this function that checked
+  // whether the passed in fork was known by the user and whether the fork had
+  // already been deleted. These assertions are now replaced with nested if
+  // statements to ensure this function is idempotent. This makes it safe to
+  // retry RRefUserDelete messages.
   {
     std::lock_guard<std::mutex> lock(mutex_);
     auto rrefIter = forks_.find(rrefId);
-    TORCH_INTERNAL_ASSERT(
-        rrefIter != forks_.end(),
-        "Inconsistent states, deleting a fork before the owner knows it.");
-    auto& rrefForks = rrefIter->second;
-    auto forkIter = rrefForks.find(forkId);
-    TORCH_INTERNAL_ASSERT(
-        forkIter != rrefForks.end(),
-        "Attempt to delete a non-exist fork ",
-        forkId);
-
-    rrefForks.erase(forkId);
-
-    if (rrefForks.empty()) {
-      auto ownerIter = owners_.find(rrefId);
-      if (ownerIter != owners_.end()) {
-        deletedRRef = ownerIter->second;
-        owners_.erase(ownerIter);
-        ownerReduced = true;
+    if (rrefIter != forks_.end()) {
+      auto& rrefForks = rrefIter->second;
+      auto forkIter = rrefForks.find(forkId);
+      if (forkIter != rrefForks.end()) {
+        rrefForks.erase(forkId);
+      } else {
+        LOG(INFO) << "Could not find UserRRef instance, "
+                  << "RRefId = " << rrefId << ", ForkId = " << forkId;
+      }
+      if (rrefForks.empty()) {
+        auto ownerIter = owners_.find(rrefId);
+        if (ownerIter != owners_.end()) {
+          deletedRRef = ownerIter->second;
+          owners_.erase(ownerIter);
+          ownerReduced = true;
+        }
+        forks_.erase(rrefIter);
       }
-      forks_.erase(rrefIter);
+    } else {
+      LOG(INFO) << "Could not find OwnerRRef with RRefId = " << rrefId;
```

Comment:
Let's add "likely because it was deleted by a previously retried message" to these two logs.